### ISSUE TITLE
Tracking: F# Hints

### ISSUE THREAD

**Issue by psfinaki** [dotnet/fsharp#14157]
> **Note**
> This is a "meta" issue for tracking purposes since GitHub does not support special types of issues.

We'll be introducing a feature called F# hints. It is inspired by C# [inline hints](https://learn.microsoft.com/en-us/visualstudio/ide/reference/options-text-editor-csharp-advanced?view=vs-2019#inline-hints) but will do more! The feature is going to be built on top of the dedicated [Roslyn API](https://github.com/dotnet/roslyn/tree/main/src/Tools/ExternalAccess/FSharp/Internal/InlineHints).

This will supersede the experimental CodeLens feature that is current present in Visual Studio. The CodeLens brand in F# is planned to do the same as in other languages, i.e. show references, test coverage and so on (details [here](https://github.com/dotnet/fsharp/issues/14036)).

The plan is to gradually add different hints and make them configurable. Here is the proposed roadmap:
- Implement inline type hints (experimental)
- Implement inline parameter name hints (experimental)
- Remove CodeLens in its current form
- Implement click-to-show behavior for the hints
- Move out of the experimental mode
- Implement signature hints
- Implement pipeline hints

**Comment by kerams** [comment]
Can hints themselves have tooltips? This here looks like an empty one:

![image](https://user-images.githubusercontent.com/5063478/200386527-a6c90aa9-8a51-488c-88ce-ccf0aa1f6b3d.png)

It would be nice to reuse type annotation tooltips with all the good stuff like clickable references:

![image](https://user-images.githubusercontent.com/5063478/200386948-1440be24-ccb7-4534-a39e-2472fb0e3ffc.png)

**Comment by vzarytovskii** [comment]
I think it's a great idea, we should make hints interactive if we can.

**Comment by psfinaki** [comment]
> Sorry if this is not the place for a Hints discussion

This IS a place for hints discussion :)

> Can hints themselves have tooltips?

I think so. As long the information doesn't aggressively jump on the user, it's only helpful.

> Also I've noticed that GetInlineHintsAsync is invoked twice after one keystroke. Is it a Roslyn thing?

On the one hand, it sounds suspicious, on the other hand, I don't have an idea what in our code can trigger this, we have a very thin layer now. I will investigate.

**Comment by majocha** [comment]
This is probably bug on Roslyn side, but hints won't go away when you release Alt+F1, unlike in C# editor.

**Comment by vzarytovskii** [comment]
> This is probably bug on Roslyn side, but hints won't go away when you release Alt+F1, unlike in C# editor.

No, it's not implemented yet. Tracked in https://github.com/dotnet/fsharp/issues/14260

**Comment by raymens** [comment]
It would be nice if a `Numerics.BigInteger` would be shown as `bigint` instead.
Not sure if there are any other similar types.
Is there already some sort of mapper to simplify names? In that case I could maybe have a go at it.

![image](https://github.com/dotnet/fsharp/assets/1369454/48a10bb6-fb4e-4e62-9976-0f285db722e6)

**Comment by psfinaki** [comment]
@raymens so, that's an interesting thing and it also applies to the tooltips. You get the full type name there unless you have specified in the opens. Take a look:

https://github.com/dotnet/fsharp/assets/5451366/64ddec02-85d6-4bd1-8e26-7bfbc8733663

This somewhat makes sense to me. Also this means that probably in many cases you'd have the short types there. WDYT?

---

That said, the tooltips for the hints themselves show the short type anyway:

![image](https://github.com/dotnet/fsharp/assets/5451366/add00d97-6df7-455b-beb6-e8236f66e0e0)

This is inconsistent and worth fixing.

---

Now, as for the mechanics of that, I think this is not a big deal and basically just depends on what info you take from the symbol:
![image](https://github.com/dotnet/fsharp/assets/5451366/ee1919ef-b580-46d5-b1c4-82989dc2a2fb)

**Comment by kerams** [comment]
I think his point is that `BigInteger` should behave like `Int32` does, where we don't use the .NET name but the F# one.

**Comment by psfinaki** [comment]
Oh, that thing. Well so the docs above pretty much answers the question :)

I am on a fence then. We can technically hardcode this conversion in the code but, given that this seems to be the only thing of that nature - and it's not very common I guess - not sure if it's worth it.

**Comment by raymens** [comment]
Thanks for the detailed explanation.
I was indeed mostly referring to the (to us highly used) `bigint` abbreviation.

Maybe the solution is to consider bigint a basic type instead... considering it's the only one apparently. However I'm going off-topic and there's probably reasons why that's the case.

**Comment by psfinaki** [comment]
Well, that's because it's arguably not that basic :)

I think, yeah, just mapping this type to its short form would be sufficient. I would be up for having it. Feel free to jump on this, or we can create a ticket and someone will eventually pick it up.
